src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -868,98 +1128,1238 @@ grpc_subchannel_call *grpc_client_channel_get_subchannel_call(
+  channel_data *chand = (channel_data *)elem->channel_data;
+  call_data *calld = (call_data *)elem->call_data;
+  if (GRPC_TRACER_ON(grpc_client_channel_trace)) {
+    char *op_str = grpc_transport_stream_op_batch_string(&batch_data->batch);
+    gpr_log(GPR_DEBUG, "chand=%p calld=%p: got on_complete, error=%s, batch=%s",
+            chand, calld, grpc_error_string(error), op_str);
+    gpr_free(op_str);
+  }
+  // Update bookkeeping in retry_state.
+  subchannel_call_retry_state *retry_state = (subchannel_call_retry_state *)
+      grpc_connected_subchannel_call_get_parent_data(
+          batch_data->subchannel_call);
+  if (batch_data->batch.send_initial_metadata) {
+    retry_state->completed_send_initial_metadata = true;
+  }
+  if (batch_data->batch.send_message) {
+    ++retry_state->completed_send_message_count;
+  }
+  if (batch_data->batch.send_trailing_metadata) {
+    retry_state->completed_send_trailing_metadata = true;
+  }
+  if (batch_data->batch.recv_initial_metadata) {
+    retry_state->completed_recv_initial_metadata = true;
+  }
+  if (batch_data->batch.recv_message) {
+    ++retry_state->completed_recv_message_count;
+  }
+  if (batch_data->batch.recv_trailing_metadata) {
+    retry_state->completed_recv_trailing_metadata = true;
+  }
+  // Check if the call is finished, and if so, get its status.
+  // The call is finished if either (a) this callback was invoked with
+  // an error or (b) we receive status.
+  bool call_finished = false;
+  grpc_status_code status = GRPC_STATUS_OK;
+  grpc_mdelem *server_pushback_md = NULL;
+  if (error != GRPC_ERROR_NONE) {  // Case (a).
+    call_finished = true;
+    grpc_error_get_status(exec_ctx, error, calld->deadline, &status, NULL,
+                          NULL);
+  } else if (batch_data->batch.recv_trailing_metadata) {  // Case (b).
+    call_finished = true;
+    grpc_metadata_batch *md_batch =
+        batch_data->batch.payload->recv_trailing_metadata
+            .recv_trailing_metadata;
+    GPR_ASSERT(md_batch->idx.named.grpc_status != NULL);
+    status = grpc_get_status_from_metadata(md_batch->idx.named.grpc_status->md);
+    if (md_batch->idx.named.grpc_retry_pushback_ms != NULL) {
+      server_pushback_md = &md_batch->idx.named.grpc_retry_pushback_ms->md;
+    }
+  }
+  if (call_finished && GRPC_TRACER_ON(grpc_client_channel_trace)) {
+    gpr_log(GPR_DEBUG, "chand=%p calld=%p: call finished, status=%s", chand,
+            calld, grpc_status_string(status));
+  }
+  // If the call is finished, check if we should retry.
+  if (call_finished &&
+      maybe_retry(exec_ctx, elem, batch_data, status, server_pushback_md)) {
+    batch_data_unref(exec_ctx, batch_data);
+    // Unref batch_data for deferred recv_initial_metadata_ready or
+    // recv_message_ready callbacks, if any.
+    if (batch_data->batch.recv_trailing_metadata &&
+        retry_state->recv_initial_metadata_ready_deferred) {
+      batch_data_unref(exec_ctx, batch_data);
```

Comment:
How is it safe to call this potentially twice in a row? Wont if unref everything in it, not just the recv_trailing_metadata